File: Assignment/BST.c
#include<stdio.h>
#include<stdlib.h>
#include <windows.h>
// #include<time.h>

typedef struct Node* node;

struct Node {
    int data;
    node left;
    node right;
};

node createNode(int data){
    node newNode = (node)malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->left = newNode->right = NULL;
    return newNode;
}

node insert(node root, int data){
    if(root == NULL){
        return createNode(data);
    }
    if(data < root->data){
        root->left = insert(root->left,data);
    }
    else if(data > root->data){
        root->right = insert(root->right,data);
    }
    return root;
}

int search(node root, int key){
    if(root == NULL){
        return 0;
    }
    if(root->data == key){
        return 1;
    }
    else if(root->data > key){
        return search(root->left,key);
    }
    else if(root->data < key){
        return search(root->right,key);
    }
    return 0;
}

void inorder(node root){
    if(root == NULL){
        return;
    }
    inorder(root->left);
    printf("%d ", root->data);
    inorder(root->right);
}

int minValue(node root){
    while(root->left != NULL){
        root = root->left;
    }
    return root->data;
}

node delete(node root, int val){
    if(root->data > val){
        root->left = delete(root->left, val);
    }else if(root->data < val){
        root->right = delete(root->right, val);
    }else{
        if(root->left == NULL && root->right == NULL){
            return NULL;
        }

        if(root->left == NULL){
            return root->right;
        }
        else if(root->right == NULL){
            return root->left;
        }
        root->data = minValue(root->right);
        root->right = delete(root->right, root->data); 
    }
    return root;
}

node insertFromFile(const char* filename) {
    FILE* file = fopen(filename, "r");
    if (file == NULL) {
        printf("Could not open file");
        return NULL;
    }
    node root = NULL;
    int num;
    while (fscanf(file, "%d", &num) == 1) {
        root = insert(root, num);
    }
    fclose(file);
    return root;
}

int main() {
    const char* filename = "case3b.txt";
    node root = insertFromFile(filename);

    printf("Inorder Traversal: ");
    inorder(root);

    int key = 25;
    LARGE_INTEGER start, end, frequency;

    QueryPerformanceFrequency(&frequency);

    QueryPerformanceCounter(&start);
    int found = search(root, key);
    QueryPerformanceCounter(&end);

    if (found) {
        printf("\nKey found");
    } else {
        printf("\nKey not found");
    }
    
    double time_taken = (double)(end.QuadPart - start.QuadPart) * 1e6 / frequency.QuadPart;
    printf("\nTime taken for searching: %lf microseconds\n", time_taken);

    // delete(root,25);
    // inorder(root);

    return 0;
}

// int main() {
//      const char* filename = "case1.txt";
//      node root = insertFromFile(filename);

//      printf("Inorder Traversal: ");
//      inorder(root);

//      int key = 25;

//      struct timespec start, end;

//      // Get start time
//      clock_gettime(CLOCK_MONOTONIC, &start);

//      // Perform search
//      int found = search(root, key);

//      // Get end time
//      clock_gettime(CLOCK_MONOTONIC, &end);

//      if (found) {
//          printf("\nKey found");
//      } else {
//          printf("\nKey not found");
//      }

//      double time_taken = (end.tv_sec - start.tv_sec) * 1e6 + (end.tv_nsec - start.tv_nsec) / 1e3;
//      printf("\nTime taken for searching: %lf microseconds\n", time_taken);

//      return 0;
// }

File: ADS/Trees/BST.java
class BST{
    static class Node{
        int data;
        Node left;
        Node right;

        Node(int data){
            this.data = data;
            this.left = this.right = null;
        }
    }

    public static Node insert(Node root, int val){
        if(root == null){
            return new Node(val);
        }
        if(root.data > val){
            root.left = insert(root.left, val);
        }
        if(root.data < val){
            root.right = insert(root.right, val);
        }
        return root;
    }

    public static void preorder(Node root){
        if(root == null){
            return;
        }
        System.out.print(root.data + " ");
        preorder(root.left);
        preorder(root.right);
    }

    public static void inorder(Node root){
        if(root == null){
            return;
        }
        inorder(root.left);
        System.out.print(root.data + " ");
        inorder(root.right);
    }

    public static void postorder(Node root){
        if(root == null){
            return;
        }
        postorder(root.left);
        postorder(root.right);
        System.out.print(root.data + " ");
    }

    public static boolean search(Node root, int key){
        if(root == null){
            return false;
        }
        if(root.data == key){
            return true;
        }
        if(root.data > key){
            return search(root.left, key);
        }
        if(root.data < key){
            return search(root.right, key);
        }
        return false;
    }

    public static Node delete(Node root, int val){
        if(root.data > val){
            root.left = delete(root.left, val);
        }
        else if(root.data < val){
            root.right = delete(root.right, val);
        }else{ // root.data == val

            // case-1 : no child
            if(root.left == null && root.right == null){
                return null;
            }

            // case-2 : one child
            if(root.left == null){
                return root.right;
            }else if(root.right == null){
                return root.left;
            }

            // case-3 : two child
            root.data = minValue(root.right);
            root.right = delete(root.right, root.data);
        }
        return root;
    }

    public static int minValue(Node root){
        while(root.left != null){
            root = root.left;
        }
        return root.data;
    }

    public static void main(String args[]){
        BST t = new BST();
        Node root = null;
        int values[] = {8,5,3,1,4,6,10,11,14};
        for(int i=0; i<values.length; i++){
            root = insert(root,values[i]);
        }
        // root = insert(root,10);
        // root = insert(root,20);
        // preorder(root)
        inorder(root);
        delete(root,5);
        System.out.println();
        inorder(root);
        // postorder(root);
        if(search(root,4)){
            System.out.println("\nKey found");
        }else{
            System.out.println("\nKey not found");
        }
    }
}